Blockbase: Add post-meta template part

mikachan opened the pull request:
<!-- Thanks for contributing to our free themes! Please provide as much information as possible with your Pull Request by filling out the following - this helps make reviewing much quicker! -->

#### Changes proposed in this Pull Request:
This PR adds a post-meta template part to Blockbase.

![Screenshot 2021-09-09 at 12 26 52](https://user-images.githubusercontent.com/1645628/132677668-47dd3bf4-80d5-42fd-a2d4-61494784c545.png)

I've mostly copied the Seedlet Blocks, Mayland Blocks, and Skatepark post-meta. I've used the 'tiny' font size; I did try 'small' as well but I thought 'tiny' probably looked better overall.

@kjellr, I know you mentioned in the original issue that we could maybe just include the post date to start with. Let me know if you think we should use different content - for now, I've gone with: author, date, categories, tags.

We're missing the correct 'post-tag' SVG icon (I think) - I've used an outline version of the 'offer' Material icon, but I think the line thickness is slightly different from the other icons. I'm not sure how we usually create the SVGs, so just pointing out that this one probably needs creating/replacing!

#### Related issue(s):
Closes #4522.

mikachan committed:
Blockbase: add post-meta template part

mikachan committed:
Blockbase: add post-tag icon

MaggieCabrera commented:
Blockbase was initially a clone of Blank Canvas, which is a child of Seedlet with a few changes, so I think copying what's on Seedlet blocks should be ok, unless Kjell thinks otherwise

mikachan committed:
Blockbase: remove tags from post meta

mikachan commented:
> Blockbase was initially a clone of Blank Canvas, which is a child of Seedlet with a few changes, so I think copying what's on Seedlet blocks should be ok, unless Kjell thinks otherwise

Ok! I've removed the tags as this then matches Seedlet Blocks exactly, but happy to adjust either way.

kjellr commented:
I think including tags is fine (I'm not sure why Seedlet Blocks doesn't display them? [The original Seedlet does](https://cloudup.com/c4nKFyhe5jc), we just don't use tags on our demo site). You can source the icon [straight from Seedlet](https://github.com/Automattic/themes/blob/1aa5058bcff44951877c2db4f461fd4a46433613/seedlet/classes/class-seedlet-svg-icons.php#L102). 

I'd also remove the separator:

Before|After
---|---
<img width="650" alt="Screen Shot 2021-09-13 at 7 59 28 AM" src="https://user-images.githubusercontent.com/1202812/133079663-56bd17e1-ba7c-464d-86c3-28b4fbe31a21.png">|<img width="636" alt="Screen Shot 2021-09-13 at 7 58 59 AM" src="https://user-images.githubusercontent.com/1202812/133079589-bbbfb426-6a2c-421f-8e78-2fd38e73e757.png">

Also, in my testing, the comments are showing up really close to this new meta area. Not sure if that's just a bug with my test site? 

<img width="670" alt="Screen Shot 2021-09-13 at 8 00 29 AM" src="https://user-images.githubusercontent.com/1202812/133079806-b915bdab-e195-4258-9800-105ffb3dec7d.png">

mikachan committed:
Blockbase: add post tags to post meta

mikachan committed:
Blockbase: add post tag SVG from Seedlet

mikachan committed:
Blockbase: remove separator from post meta

mikachan commented:
Thanks @kjellr! I've added the post tags back in, copied the post tag SVG over from Seedlet (this looks great, thanks for pointing me in the right direction), and removed the separator.

Seedlet Blocks doesn't include the tags in the post meta template. I could add them here too while I'm working in this area, if you think they should be there?

The comments aren't really close on my local dev, looks like they're inheriting the top margin from this selector: `.wp-site-blocks > * + *` - 

![Screenshot 2021-09-13 at 14 21 52](https://user-images.githubusercontent.com/1645628/133091436-3908d2f7-55c9-4e8e-a9ed-7f2421c05bc3.png)

They're using the `--wp--style--block-gap` variable, so I don't know if this is related to a Gutenberg update, or maybe my local version is out of date.. Will have a play with this now.

kjellr commented:
> Seedlet Blocks doesn't include the tags in the post meta template. I could add them here too while I'm working in this area, if you think they should be there?

Yeah, I think they should be there. Feel free to do it here if you'd like, but it could totally be a separate PR if that's easier.

kjellr commented:
> The comments aren't really close on my local dev, looks like they're inheriting the top margin from this selector: .wp-site-blocks > * + * -
> They're using the --wp--style--block-gap variable, so I don't know if this is related to a Gutenberg update, or maybe my local version is out of date.. Will have a play with this now.

If we could make this at least `calc(2 * var(--wp--style--block-gap))`, that would be great.

mikachan committed:
Blockbase: add spacer above post comments block

mikachan committed:
Seedlet Blocks: add post tags to post meta

mikachan commented:
I've updated Seedlet Blocks so the post meta now includes the tags:

![Screenshot 2021-09-13 at 16 58 16](https://user-images.githubusercontent.com/1645628/133117198-e28ee695-c920-4f92-9f80-af50a2855c76.png)

And I managed to replicate the lack of spacing above the post comments block (my Gutenberg was out of date). I've added a spacer block above the post comments on Blockbase that's the same height as `calc(2 * var(--wp--style--block-gap))`. I added it as a spacer as all the child themes use a spacer block above the comments, and it also means we don't need to change the CSS for Blockbase and all the children.

![Screenshot 2021-09-13 at 17 00 47](https://user-images.githubusercontent.com/1645628/133117659-b3d84b0b-bca8-42ce-b400-8935455df64e.png)

This is ready for another review, thanks!

kjellr commented:
Thanks, @mikachan! Just one minor comment: Can we increase the gap between items to `calc(2 * var(--wp--custom--margin--baseline))` in Blockbase? This helps it match the gap size that Seedlet & Blank Canvas use normally. 

Before|After
---|---
<img width="360" alt="Screen Shot 2021-09-14 at 8 08 56 AM" src="https://user-images.githubusercontent.com/1202812/133255113-647a56d1-386a-4007-ab81-867db1c70324.png">|<img width="386" alt="Screen Shot 2021-09-14 at 8 17 41 AM" src="https://user-images.githubusercontent.com/1202812/133256097-ee5af875-1e14-4995-99ca-6a1cc34bb603.png">


After that, I think this is all set for now!

kjellr commented:
> Can we increase the gap between items...

For what it's worth, I think the theme needs to opt into [the new gap feature](https://href.li/?https://github.com/WordPress/gutenberg/pull/33991) for that to work? I'm not sure how to actually set the gap for the block though. 🤔

danieldudzic commented:
> > Can we increase the gap between items...
> 
> For what it's worth, I think the theme needs to opt into [the new gap feature](https://href.li/?https://github.com/WordPress/gutenberg/pull/33991) for that to work? I'm not sure how to actually set the gap for the block though. 🤔

So, I have checked the following blocks: `post-author`, `post-date ` and `post-terms` - and they seem to be missing support for `gap` in the core.

If my understanding is correct, gap makes an obvious sense with blocks that will be used in multiple instances together, like `buttons` or `columns`.

With `post-author`, `post-date` and `post-terms` being individual blocks I can see how `gap` might be tricky.

It would almost make sense to have just one block called post meta, and then have different parts inside of the block being `post-author`, `post-date` and `post-terms`. That would make `gap` super useful.

kjellr commented:
I _think_ you only need gap support on the Group block, right? Since that's the one that has the `flex` style applied to it?

mikachan committed:
Blockbase: enable blockGap & set gap size for all children

mikachan commented:
> For what it's worth, I think the theme needs to opt into the new gap feature for that to work?

Thank you! I tried with spacers and then resorted to CSS... but setting `blockGap` to true in theme.json seems to have done the trick. 

I've set the Blockbase gap size to `calc(2 * var(--wp--custom--margin--baseline))`, and then set all the children gap sizes to `var(--wp--custom--gap--vertical)`, as this was the previous value in theme.json. However, `blockGap` wasn't previously enabled, so all the themes were using the default gap size of `0.5em` (which comes from Gutenberg I guess). This means all gap sizes are bigger now - would it be better to set them all to the old default value so there's no change?

Also, I could only seem to get `blockGap` working if it was applied globally (`styles.spacing.blockGap`), and I couldn't get it to work if I tried to set it directly on a block (e.g. `styles.blocks.core/post-meta.spacing.blockGap`). Or I was doing it wrong!

mikachan committed:
Merge branch 'trunk' into add/blockbase-post-meta

mikachan committed:
Blockbase children: sync up child-theme.json files

They were all using --wp--custom--margin--vertical instead of --wp--custom--gap--vertical

kjellr commented:
> However, blockGap wasn't previously enabled, so all the themes were using the default gap size of 0.5em (which comes from Gutenberg I guess). This means all gap sizes are bigger now - would it be better to set them all to the old default value so there's no change?

I _think_ it probably makes sense to keep the default size the same for now. I'm not sure how many places that gap property is used, so I think a more conservative approach is warranted here.

mikachan committed:
Blockbase children: change gap size to GB default of 0.5em

mikachan commented:
Thanks! I've set all of the Blockbase children gap sizes to `0.5em`. Ready for another review.

scruffian commented:
This now looking a bit odd on Seedlet: 
<img width="431" alt="Screenshot 2021-09-15 at 18 51 04" src="https://user-images.githubusercontent.com/275961/133484191-06c72d39-0f52-40e5-b34b-904383e5e4a9.png">

Could all the children just inherit from Blockbase?

mikachan committed:
Blockbase & co: use gap baseline across BB & all children

mikachan commented:
> Could all the children just inherit from Blockbase?

Yeah I think they can. I didn't do this originally as I only tried `var(--wp--custom--gap--vertical)` globally, which looked too big on most themes. I've just changed all themes to use `calc(2 * var(--wp--custom--gap--baseline))`, the same as Blockbase, and this seems to work well across them all. I'm not sure why I didn't try this before, seems like the obvious choice out of the gap sizes! Here's what this is looking like after this change:

Blockbase:

![image](https://user-images.githubusercontent.com/1645628/133598625-ffd196d2-975f-4703-9c58-feb7b2a0320b.png)

Seedlet Blocks:

![image](https://user-images.githubusercontent.com/1645628/133598710-d9229aac-11fb-4a42-b7b4-fc8342de96ff.png)

Mayland Blocks:

![image](https://user-images.githubusercontent.com/1645628/133598770-343afe32-60d7-4601-9b42-2afa3f6db68b.png)

Quadrat:

![image](https://user-images.githubusercontent.com/1645628/133598819-390c7a6a-742d-4238-ab92-262e328bdb94.png)

(This doesn't affect Skatepark and Geologist.)

kjellr commented:
This looks great. One last question: Is it possible to specify a different vertical/horizontal gaps? 

If these meta items wrap, it uses the same horizontal gap for the vertical spacing: 

<img width="521" alt="Screen Shot 2021-09-16 at 7 44 37 AM" src="https://user-images.githubusercontent.com/1202812/133606476-ff0799ef-f638-44a3-86c1-3ecc3fd75c93.png">

I'd prefer if the vertical spacing was half of that: 

<img width="521" alt="Screen Shot 2021-09-16 at 7 44 37 AM" src="https://user-images.githubusercontent.com/1202812/133606612-4cba1cf6-ae8b-4b6d-8fc2-45b4b17e311a.png">

If that's not possible yet, it's ok, and we can merge this as-is for now. Thank you!

mikachan commented:
> Is it possible to specify a different vertical/horizontal gaps?

From what I can tell, the new blockGap feature only supports one value, and can't yet split out different horizontal and vertical gaps. It looks like there's an ongoing issue for it though: https://github.com/WordPress/gutenberg/issues/34529 (and there's an issue from @scruffian here too: https://github.com/WordPress/gutenberg/issues/34347)

We could set the vertical gap for the post-meta with CSS until this feature is available? Although it looks like it's tricky to overwrite the specific `.wp-container-ID` styles, so it might have to be something like:

```
.post-meta {
	row-gap: var(--wp--custom--gap--baseline) !important;
}
```

kjellr commented:
Maybe we could set it via CSS just for the post meta?

mikachan committed:
Blockbase: add row-gap value to post-meta

mikachan committed:
Skatepark: remove gap value

This is now applied via theme.json (as the same value)

mikachan commented:
Yeah I think this works, this is how it looks:

Blockbase:

![image](https://user-images.githubusercontent.com/1645628/133639424-9b2477f3-f486-4974-a184-59ac1376107f.png)

Seedlet Blocks:

![image](https://user-images.githubusercontent.com/1645628/133639513-bcda5f17-3e7e-40b0-9d34-1c722d36a919.png)

Mayland Blocks:

![image](https://user-images.githubusercontent.com/1645628/133639602-028b743f-9024-4790-801a-02aad5138281.png)

kjellr commented on blockbase/sass/post/_meta.scss:
```diff
@@ -1,4 +1,6 @@
 .post-meta {
+	row-gap: var(--wp--custom--gap--baseline) !important;
+
```
Might make sense to add a code comment here to explain why we're using `!important`. And we can also probably remove that empty line below.

kjellr reviewed (approved):
Left one minor comment, but otherwise I think this is good to go. Thank you! 🙌

mikachan committed:
Blockbase: add comment next to use of !important

scruffian commented on blockbase/sass/post/_meta.scss:
```diff
@@ -1,4 +1,5 @@
 .post-meta {
+	row-gap: var(--wp--custom--gap--baseline) !important; // !important is needed to override the unique .wp-container-ID classes
```
It would be good to link the GB issue that suggests different vertical and horizontal block gaps, so we can remove this when that gets fixed.

mikachan replied:
Ah yeah good idea, will sort this now

mikachan committed:
Blockbase: add link to GB issue around block gaps

scruffian reviewed (approved):
LGTM

mikachan committed:
Merge branch 'trunk' into add/blockbase-post-meta
